### DISCUSSION TITLE
Change graph according to configuration

### DISCUSSION THREAD

**Discussion by jondoe69420** [dagster-io/dagster#21215] [Q&A] [answered]
Hi everyone.

Is there a way in dagster to change/create a graph according to a runtime configuration?

I have a use-case where I have multiple different ops defined, but I want to run a different subset of them every time I run the pipeline, and I want the order to be configurable.

For example, if I have defined Op A, Op B, and Op C, I want to be able to dynamically run the following

A -> B -> C
B -> A -> C
C -> B -> A
A -> C

and so on.

Is this possible currently in dagster?

**Comment by garethbrickman** [comment, marked as answer]
Dagster does not support fully dynamic reordering of ops or creation of new graph structures based solely on runtime configuration. The structure of a graph is defined at build time, not at runtime. This means that the shape of the graph, including the ops and the dependencies between them, must be known before the graph is executed. However, there are some ways to achieve dynamic behavior within certain constraints.

One approach to handle dynamic execution paths is to use [conditional branching within your graph](https://docs.dagster.io/concepts/ops-jobs-graphs/graphs#with-conditional-branching). You can model conditional execution by taking advantage of the fact that an op only starts to execute once all its inputs have been resolved. This allows you to create branches in your graph where only certain paths are taken based on runtime data.

For example, you could have an op that decides which path to take, and downstream ops would only execute if their inputs are resolved, effectively creating a dynamic flow. However, this is not the same as dynamically reordering ops or creating entirely new graph structures at runtime.

Another approach mentioned is to use dynamic orchestration with the `map` operation, which allows for dynamic generation of execution steps based on runtime data. This could be used to process a set of inputs dynamically, but it doesn't provide a way to dynamically reorder the entire graph structure.

If you need to switch between entirely different graph structures based on runtime configuration, one potential solution is to define multiple graphs or jobs that represent each of the desired execution orders and then select which one to execute at runtime. This selection could be based on a configuration value or some other runtime data.

For example, you could define separate jobs for each of the execution orders you mentioned:

```python
@job
def job_a_b_c():
    c_op(b_op(a_op()))

@job
def job_b_a_c():
    c_op(a_op(b_op()))

@job
def job_c_b_a():
    a_op(b_op(c_op()))

@job
def job_a_c():
    c_op(a_op())
```

Then, at runtime, you could select which job to execute based on your configuration or criteria. However, this approach requires you to define all possible permutations ahead of time, which may not be practical for a large number of ops or complex dependency structures.

**Reply by jondoe69420** [reply]
@garethbrickman Thanks for the detailed answer! Is this perhaps a feature that could be supported in the future? Or is this completely out of scope?

**Reply by garethbrickman** [reply]
@jondoe69420 I don't think anything like it is being actively worked on or planned. We have [an Ideas section in our Discussions](https://github.com/dagster-io/dagster/discussions/new?category=ideas) if you want to let the Community know about it!
